src/StructuredLogger/BinaryLogger/BuildEventArgsReader.cs

```diff
@@ -203,10 +224,91 @@ private static bool IsAuxiliaryRecord(BinaryLogRecordKind recordKind)
         private void ReadBlob(BinaryLogRecordKind kind)

```

Comment:
In this entire ReadBlob() method we should be very mindful of allocations, and be explicit about every allocation happening. I'm dealing with a binlog that is 900 MB in size, 600 MB of which are the embedded .nuget.dgspec.json files (https://github.com/NuGet/Home/issues/12957). They expand to over 10 GB in memory.

I haven't read this closely, but let's make sure every allocation is accounted for. Ideally I also wanted to make ArchiveFiles lazy if possible, so that they decompress their file content only when requested, to save on memory. But that's for later. For now let's just double-check that this only allocates where necessary and avoids unnecessary allocations.

Reply:
https://github.com/KirillOsenkov/MSBuildStructuredLog/pull/711#discussion_r1384920463 applies here as well :-)

Basically - the existing (reading) workflow should be untouched.
However it still means that during reading the entire embedded archive is greedily decompressed and transformed to files in memory. This can definitely be improved. Ideally with one more binlog format breaking change (but that's for a separate talk)